### PR TITLE
Selecting a pawn in the last row crashes the game

### PR DESCRIPTION
**Pull request by Kinrany**
There's a bug where selecting a pawn that reached the last row crashes the game due to referencing cells outside the board.

This PR:

 * adds an `onBoard(n: number): boolean` function that replaces all the `x >= 0 && x < 8` checks
 * in `pawnLogic`, changes the `forward(y, n)` function to always use `pawn.y` as it's first argument
 * fixes the bug
 * changes move order in `pawnLogic` so that moving two squares forward is the last

Probably easier to review commit by commit.

### CODE REVIEW

File: demo/src/components/logic.ts [hunk 1]
```diff
@@ -1,32 +1,41 @@
...
+                let enemy = lookup(x, forward(1));
+                if (enemy !== undefined && enemy.color !== pawn.color) {
+                    result.push([x, forward(1)]);
+                }
             }
         }
     }
+    if (
+        !pawn.wasMoved &&
+        onBoard(forward(2)) &&
```
Comment (jvanbruegge) [inline demo/src/components/logic.ts]: This check is actually not needed, because if the pawn was not moved we know we are on board

**Review by jvanbruegge — COMMENTED** [review]
Very good changes! Thanks.

File: demo/src/components/logic.ts (hunk 1 above)
Comment (Kinrany) [reply demo/src/components/logic.ts]: Yeah, wasn't sure about it.

Relying on initial board configuration feels dirty. On the other hand, rules do not generalize trivially. Like, should it be `pawn.wasMoved`, or `pawn.color === 'white' && pawn.x === 2 || pawn.color === 'black' && pawn.x === 7`, or both?

I'll remove it.

**Comment by Kinrany** [comment]
Note to self: spell check before saving text to an immutable store

**Comment by jvanbruegge** [comment]
`git commit --amend` lets you redo your commit

**Comment by Kinrany** [comment]
Both versions will show up here anyway, I think?

**Comment by Kinrany** [comment]
Oh, they won't. TIL, thanks :)

**Comment by jvanbruegge** [comment]
If you need to change more, you can use `git rebase -i HEAD~5` which lets xou change the last 5 (in this case) commits

**Comment by Kinrany** [comment]
Yeah, I just didn't know that Github updates the commit messages in issues/PRs.

Shall I squash all ~~four~~ three?

**Comment by jvanbruegge** [comment]
No, i like the distictions

**Pull request merged**